utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java

```diff
@@ -199,17 +266,291 @@ public <T> Builder put(Key<T> key, T value) {
+                    Value<?> newValue = v.copy();
+                    valueRemapping.put(v, newValue);
+                    attributes.put(k, newValue);
+                });
+                dependencyGraph = dependencyGraph.copy(valueRemapping::get);
+                copyOnUpdate = false;
+            }
+        }
+
         @Override
         public AttributeMap build() {
-            return new AttributeMap(configuration);
+            AttributeMap result = new AttributeMap(this);
+            copyOnUpdate = true;
+            return result;
+        }
+    }
+
+    /**
+     * A value that is evaluated lazily. See {@link Builder#putLazy(Key, LazyValue)}.
+     */
+    @FunctionalInterface
+    public interface LazyValue<T> {
+        T get(LazyValueSource source);
+    }
+
+    /**
+     * A source for other values, provided to a {@link LazyValue} when the value is resolved.
+     */
+    @FunctionalInterface
+    public interface LazyValueSource {
+        <T> T get(Key<T> sourceKey);
+    }
+
+    /**
+     * Tracks which values "depend on" other values, so that when we update one value, when can clear the cache of any other
+     * values that were derived from the value that was updated.
+     */
+    private static final class DependencyGraph {
+        /**
+         * Inverted adjacency list of dependencies between derived keys. Mapping from a key to what depends on that key.
+         */
+        private final Map<Key<?>, Set<Value<?>>> dependents;
+
+        private DependencyGraph() {
+            this.dependents = new HashMap<>();
+        }
+
+        private DependencyGraph(Map<Key<?>, Set<Value<?>>> dependents,
+                                Function<Value<?>, Value<?>> valueRemapping) {
+            this.dependents = new HashMap<>(dependents.size());
+            dependents.forEach((key, values) -> {
+                Set<Value<?>> newValues = new HashSet<>(values.size());
+                this.dependents.put(key, newValues);
+                values.forEach(v -> newValues.add(valueRemapping.apply(v)));
+            });
+        }
+
+        private void addDependency(Value<?> consumer, Key<?> dependency) {
+            dependents.computeIfAbsent(dependency, k -> new HashSet<>()).add(consumer);
+        }
+
+        private Set<Value<?>> dependents(Key<?> key) {
+            return dependents.getOrDefault(key, emptySet());
+        }
+
+        private void invalidateConsumerCaches(Key<?> key) {
+            Set<Value<?>> dependents = dependents(key);
+            if (dependents.isEmpty()) {
+                return;
+            }
+
+            Queue<Value<?>> unloadQueue = new ArrayDeque<>(dependents);
+            dependents.clear();
+            while (!unloadQueue.isEmpty()) {
+                Value<?> toUnload = unloadQueue.poll();
+
+                Set<Value<?>> toUnloadDependents = dependents(toUnload.key());
+                unloadQueue.addAll(toUnloadDependents);
+                toUnloadDependents.clear();
```

Comment:
Why does this get cleared here?

Reply:
After we have cleared the cache for a value, the things that it depends on might change when it reloads. If we didn't clear this, it means that a value's old dependencies still govern its cache, even if it might not depend on those values.

Example scenario: Pretend the retry mode depends on the defaults mode, and because the defaults mode is "in-region" it depends on the region. (RetryMode -> DefaultsMode -> Region)
1. We modify the Region, so we clear the cache for RetryMode and DefaultsMode
2. We change the DefaultsMode from in-region to standard
3. We build the attribute map, so we derive and re-cache the values for RetryMode and DefaultsMode

If we didn't do `toUnloadDependents.clear()` the dependency graph would still say that DefaultsMode depends on Region, even though we changed the defaults mode to standard and it no longer has a dependency on region. That means when we changed region, we'd still reset the cache of DefaultsMode, even though it doesn't need to be reset.

Reply:
Ah I see. I didn't consider that it could be that dynamic!